### ISSUE TITLE
Implement alternate camera projection(s)

### ISSUE THREAD

**Issue by zappedfish** [godotengine/godot-proposals#3779]
### Describe the project you are working on

- A game with a first-person view where it's _important to fit as much surroundings inside the screen as possible._
  - For example **a competitive shooter**
  - Or a game with **unique artstyle**
- Planetarium simulation

Currently I'm working on a first-person game where high FOVs are needed.

### Describe the problem or limitation you are having in your project

Standard rectilinear projection greatly distorts the surroundings at high FOVs, compared to alternative projection methods. I can't fit as much of surroundings as I'd like to.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Alternative projection methods help fit more of surroundings inside the screen **without extreme distortions**.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

An existing shader by Cykyrios https://github.com/Cykyrios/Godot360 which works by stitching together multiple cameras.

Comparison of standard rectilinear and panini projections:

![bildo](https://user-images.githubusercontent.com/36744029/148695127-445491fb-ef52-4a5e-834d-b892fab1ca37.png)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

While this can be implemented using a shader, compatibility and performance are very poor.

### Is there a reason why this should be core and not an add-on in the asset library?

Such method can introduce rendering artefacts, for example, [while using built-in effects, such as "Bloom"](https://github.com/godotengine/godot/issues/7340#issuecomment-573401411), and other screen-space effects.

To avoid issues and improve performance, such feature must be a part of the rendering pipeline. _(I assume?)_

**Comment by lentsius-bark** [comment]
I'd like to chip in to say i fully support the proposal, I have wanted to use higher FOVs in a first person project of mine that I couldn't due to the distortion which was far greater than I thought it would, and due to the project's time line I didn't have the time to look into shaders. This would be great.

**Comment by Rinocrosser** [comment]
Yep, this is very important for almost any first-person game

**Comment by Calinou** [comment]
@Rinocrosser Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by Rinocrosser** [comment]
@Calinou Hi! Is there any milestone or something?

**Comment by Calinou** [comment]
> @Calinou Hi! Is there any milestone or something?

No, as we don't know exactly how this should be implemented. It's also not a feature that is very commonly used by players in games that support it, and it can be simulated with a shader in some cases already.

**Comment by myaaaaaaaaa** [comment]
It's actually pretty easy to get this effect right now by using vertex shaders, [since handling camera projections is already one of its primary roles](https://docs.godotengine.org/en/latest/tutorials/shaders/shader_reference/spatial_shader.html#vertex-built-ins) - just ignore the default projection matrix and write your own projection transform, no need for changes to core or stitching together multiple cameras.

[You'll probably want tessellation support though](5995), since models that aren't highpoly enough will end up with lowpoly curved projections.

Things get a bit hairier once fov reaches 180, since culling will probably break. Past that, the wider you go, the more likely it is for triangles behind the camera to wrap around onto opposite ends of the screen if they're large enough, furthering the argument for tessellation.

That said, even for 180+ fov, I think vertex shaders with tessellation are the cleanest way to get alternate camera projections today, unless you'd prefer to wait for the day when Godot is rewritten as a full raytracing engine 🙃

**Comment by xPuls3** [comment]
I would like to point out that panini projection helps combat vertigo for people who experience it when playing games with a higher field of vision. As mentioned in the other issue above, lowering the FOV to 45 (horizontal, 27 vertical) or lower is required to mitigate distortion, though when paired with rapid camera movement (such as in fast paced games) having a field of vision that low is unfortunately another cause of vertigo.

Another use case and reason for implementing it, should anyone be up to it.

**Comment by VividVerdigris** [comment]
This isn't a complete replacement for having the Pannini projection as a built-in option, as it still has some issues, but I made a Pannini projection compositor effect available [here](https://codeberg.org/verdigris/pannini-projection-godot). It works in a different way to Godot360 and is more like Unreal Engine's implementation, so the effect itself has very little performance impact (but that comes with its own caveats---see the repo for more details).

**Comment by hypotensor** [comment]
> This isn't a complete replacement for having the Pannini projection as a built-in option, as it still has some issues, but I made a Pannini projection compositor effect available [here](https://codeberg.org/verdigris/pannini-projection-godot). It works in a different way to Godot360 and is more like Unreal Engine's implementation, so the effect itself has very little performance impact (but that comes with its own caveats---see the repo for more details).

Thanks for uploading this. I played around with your shader and it's quite remarkable. If you use variable rate shading and supersampling, you can achieve a remarkably fine image quality. I think the projection correction is the way to go, not the cubemap route suggested in the original post.
